elmslie-core/src/main/java/vivid/money/elmslie/core/store/ElmStore.kt

```diff
@@ -17,123 +14,75 @@ class ElmStore<Event : Any, State : Any, Effect : Any, Command : Any>(
-            childStore.effects { effectMapper(currentState, it)?.let(::dispatchEffect) },
-            childStore.states { dispatchState(stateReducer(currentState, it)) },
-        )
-        childStore.start()
         return this
     }
 
-    private fun dispatchState(state: State) = synchronized(stateLock) {
-        stateInternal.set(state)
-        stateListeners.forEach { it(state) }
+    override fun stop() {
+        _isStarted.set(false)
+        storeScope.cancel()
     }
 
-    private fun dispatchEffect(effect: Effect) {
-        logger.debug("New effect: $effect")
-        effectListeners.forEach { it(effect) }
-    }
+    override fun states(): Flow<State> = statesFlow.asStateFlow()
+
+    override fun effects(): Flow<Effect> = effectsFlow.asSharedFlow()
 
     private fun dispatchEvent(event: Event) {
-        executor.submit {
+        storeScope.launch {
             try {
                 logger.debug("New event: $event")
-                eventListeners.forEach { it(event) }
-                val result = reducer.reduce(event, currentState)
-                dispatchState(result.state)
-                result.effects.forEach(::dispatchEffect)
-                result.commands.forEach(::executeCommand)
+                val (state, effects, commands) = reducer.reduce(event, currentState)
+                statesFlow.value = state
+                effects.forEach(::dispatchEffect)
+                commands.forEach(::executeCommand)
+            } catch (error: CancellationException) {
+                throw error
             } catch (t: Throwable) {
                 logger.fatal("You must handle all errors inside reducer", t)
             }
         }
     }
 
-    private fun executeCommand(command: Command) = try {
-        logger.debug("Executing command: $command")
-        disposables += actor.execute(command, ::dispatchEvent, { logger.nonfatal(error = it) })
-    } catch (t: Throwable) {
-        logger.fatal("Unexpected actor error", t)
+    private fun dispatchEffect(effect: Effect) {
+        storeScope.launch {
+            logger.debug("New effect: $effect")
+            effectsFlow.emit(effect)
+        }
     }
 
-    private fun startBuffering() = effectListeners.add(effectBufferingListener)
-
-    private fun stopBuffering() = effectListeners.remove(effectBufferingListener)
-
-    private fun dispatchBuffer(onEffectEmission: (Effect) -> Unit) = effectBuffer
-        .onEach { onEffectEmission(it) }
-        .clear()
-
-    private fun requireNotStarted() {
-        if (!isStartedInternal.compareAndSet(false, true)) {
-            logger.fatal("Store start error", StoreAlreadyStartedException())
+    private fun executeCommand(command: Command) =
+        try {
+            logger.debug("Executing command: $command")
+            storeScope.launch {
+                actor
+                    .execute(command)
+                    .catch { logger.nonfatal(error = it) }
```

Comment:
I think that all errors will be caught here and we don't need to wrap this by try catc